src/composition.rs

```diff
@@ -927,6 +1037,73 @@ from(bucket: "my-bucket") |> yield(name: "my-result")
     |> filter(fn: (r) => r._measurement == "anMeasurement")
     |> filter(fn: (r) => r._field == "firstField" or r._field == "secondField")
     |> yield(name: "_editor_composition")
+"#
+            .to_string(),
+            composition.to_string()
+        )
+    }
+
+    #[test]
+    fn composition_add_tag_value() {
+        let fluxscript = r#"from(bucket: "an-composition")
+        |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
+        |> yield(name: "_editor_composition")
+    "#;
+        let ast = flux::parser::parse_string("".into(), &fluxscript);
+
+        let mut composition = Composition::new(ast);
+        // DON'T INITIALIZE THIS! WE'RE SIMULATING AN ALREADY INITIALIZED QUERY.
+        composition.add_tag_value(&"tagKey", &"tagValue").unwrap();
+
+        assert_eq!(
+            r#"from(bucket: "an-composition")
+    |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
+    |> filter(fn: (r) => r.tagKey == "tagValue")
+    |> yield(name: "_editor_composition")
+"#
+            .to_string(),
+            composition.to_string()
+        )
+    }
+
+    #[test]
+    fn composition_add_tag_value_tag_value_already_exists() {
+        let fluxscript = r#"from(bucket: "an-composition")
+        |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
+        |> filter(fn: (r) => r._measurement == "anMeasurement")
+        |> filter(fn: (r) => r.tagKey == "tagValue")
+        |> yield(name: "_editor_composition")
+    "#;
+        let ast = flux::parser::parse_string("".into(), &fluxscript);
+
+        let mut composition = Composition::new(ast);
+        // DON'T INITIALIZE THIS! WE'RE SIMULATING AN ALREADY INITIALIZED QUERY.
+
+        assert!(composition
+            .add_tag_value(&"tagKey", &"tagValue")
+            .is_err());
+    }
+
+    #[test]
+    fn composition_add_tag_value_second_tag_value() {
+        let fluxscript = r#"from(bucket: "an-composition")
+        |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
+        |> filter(fn: (r) => r._measurement == "anMeasurement")
+        |> filter(fn: (r) => r.tagKey1 == "tagValue1")
+        |> yield(name: "_editor_composition")
+    "#;
+        let ast = flux::parser::parse_string("".into(), &fluxscript);
+
+        let mut composition = Composition::new(ast);
+        // DON'T INITIALIZE THIS! WE'RE SIMULATING AN ALREADY INITIALIZED QUERY.
+        composition.add_tag_value(&"tagKey2", &"tagValue2").unwrap();
+
+        assert_eq!(
+            r#"from(bucket: "an-composition")
+    |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
+    |> filter(fn: (r) => r._measurement == "anMeasurement")
+    |> filter(fn: (r) => r.tagKey1 == "tagValue1" and r.tagKey2 == "tagValue2")
```

Comment:
Hm, I _think_ these need to be two separate `filter` calls, particularly because there might be something in the middle there. For instance:

```
from(...) |> range(...)
  filter(fn: (r) => r._tagKey1 == "tagValue1")
  filter(fn: (r) => r._field == "firstField" or r._field == "secondField")
  filter(fn: (r) => r._tagKey2 == "tagValue2")
```

...could have a different meaning/different query plan than...

```
from(...) |> range(...)
  filter(fn: (r) => r._tagKey1 == "tagValue1" and r._tagKey2 == "tagValue2")
  filter(fn: (r) => r._field == "firstField" or r._field == "secondField")
```

Reply:
I agree that these mean different things for the query plan. But for the goal of schema composition, we have to pick a single pattern to teach the user. (Because our schema composition injection will always be with this pattern. And the advanced user can modify post-composition sync.)

What is the common case? What will most beginning-to-early-intermediate users want?

Reply:
Per discussion, here is a PR added to the end of the PR chain. Makes the requested change:
https://github.com/influxdata/flux-lsp/pull/560